util/deadlockhistory/deadlock_history.go

```diff
@@ -73,8 +73,28 @@ func NewDeadlockHistory(capacity int) *DeadlockHistory {
 
 // GlobalDeadlockHistory is the global instance of DeadlockHistory, which is used to maintain recent several recent
 // deadlock events globally.
-// TODO: Make the capacity configurable
-var GlobalDeadlockHistory = NewDeadlockHistory(10)
+// The real size of the deadlock history table should be initialized with `Resize`
+// in `setGlobalVars` in tidb-server/main.go
+var GlobalDeadlockHistory = NewDeadlockHistory(0)
+
+// Resize update the DeadlockHistory's table max capacity to newCapacity
+func (d *DeadlockHistory) Resize(newCapacity uint) {
+	d.Lock()
+	defer d.Unlock()
+	if newCapacity != uint(len(d.deadlocks)) {
+		current := d.getAll()
+		d.head = 0
+		if uint(len(current)) < newCapacity {
+			// extend deadlocks
+			d.deadlocks = make([]*DeadlockRecord, newCapacity)
+			copy(d.deadlocks, current)
+		} else {
+			// shrink deadlocks, keep the last len(current)-newCapacity items
+			d.deadlocks = current[uint(len(current))-newCapacity:]
```

Comment:
I'm afraid after the sub-slicing, it still references the old large slice and the discarded elements may not be freed by golang. But I'm not sure. @cfzjywxk Are you familiar with this kind of behaviors of golang?

Reply:
If the underlying elements has memory reference to some other big objects, the memory may not be freed in time, for example the tidb batch client memory reset here https://github.com/pingcap/tidb/blob/master/store/tikv/client/client_batch.go#L128.